diff --git a/arch/arm/dts/sun50i-a64-pine64-plus-u-boot.dtsi b/arch/arm/dts/sun50i-a64-pine64-plus-u-boot.dtsi
index 9c61bea..fad46da 100644
--- a/arch/arm/dts/sun50i-a64-pine64-plus-u-boot.dtsi
+++ b/arch/arm/dts/sun50i-a64-pine64-plus-u-boot.dtsi
@@ -16,13 +16,13 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			pinctrl-names = "default";
-			pinctrl-0 = <&rgmii_pins>;
-			phy-mode = "rgmii";
-			phy = <&phy1>;
+			pinctrl-0 = <&rmii_pins>;
+			phy-mode = "rmii";
+			phy = <&phy0>;
 			status = "okay";
 
-			phy1: ethernet-phy@1 {
-				reg = <1>;
+			phy0: ethernet-phy@0 {
+				reg = <0>;
 			};
 		};
 	};
diff --git a/arch/arm/dts/sun50i-a64-pine64.dts b/arch/arm/dts/sun50i-a64-pine64.dts
index c680ed3..8746184 100644
--- a/arch/arm/dts/sun50i-a64-pine64.dts
+++ b/arch/arm/dts/sun50i-a64-pine64.dts
@@ -52,6 +52,7 @@
 
 	aliases {
 		serial0 = &uart0;
+		ethernet = &emac;
 	};
 
 	chosen {
@@ -64,6 +65,36 @@
 		regulator-min-microvolt = <3300000>;
 		regulator-max-microvolt = <3300000>;
 	};
+
+	soc {
+    emac: ethernet@01c30000 {
+			compatible = "allwinner,sun50i-a64-emac";
+      reg = <0x01c30000 0x2000>, <0x01c00030 0x4>;
+      reg-names = "emac", "syscon";
+      interrupts = <GIC_SPI 82 IRQ_TYPE_LEVEL_HIGH>;
+      resets = <&ccu RST_BUS_EMAC>;
+      reset-names = "ahb";
+      clocks = <&ccu CLK_BUS_EMAC>;
+      clock-names = "ahb";
+      #address-cells = <0>;
+      #size-cells = <0>;
+      pinctrl-names = "default";
+      pinctrl-0 = <&rmii_pins>;
+      phy-mode = "rmii";
+      phy = <&phy0>;
+
+      snps,reset-gpios = <&phy_reset_pins>;
+      snps,reset-active-low = <1>;
+      snps,reset-delay-us = <200>;
+
+      status = "okay";
+      
+      phy0: ethernet-phy@0 {
+        reg = <0>;
+      };
+    };
+  };
+
 };
 
 &ehci1 {
@@ -109,3 +140,15 @@
 &usbphy {
 	status = "okay";
 };
+
+&pio {
+	rmii_pins: rmii_pins {
+		allwinner,pins = "PD10", "PD11", "PD13", "PD14",
+				 "PD17", "PD18", "PD19", "PD20",
+				 "PD22", "PD23";
+		allwinner,function = "emac";
+		allwinner,drive = <3>;
+		allwinner,pull = <0>;
+	};
+};
+
diff --git a/arch/arm/dts/sun50i-a64.dtsi b/arch/arm/dts/sun50i-a64.dtsi
index 35f0c7e..940f5ef 100644
--- a/arch/arm/dts/sun50i-a64.dtsi
+++ b/arch/arm/dts/sun50i-a64.dtsi
@@ -317,11 +317,12 @@
 				pins = "PG8", "PG9";
 				function = "uart1";
 			};
-			
+
 			phy_reset_pins: phy_reset_pins {
 				pins = "PD24";
 				function = "emac";
 			};
+
 		};
 
 		uart0: serial@1c28000 {
